File: theta_star_smooth_planner/theta_star_smooth_planner.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from nav_msgs.msg import OccupancyGrid, Path
from geometry_msgs.msg import PoseStamped, Pose
from rclpy.qos import QoSProfile, DurabilityPolicy
from tf2_ros import Buffer, TransformListener, LookupException
from queue import PriorityQueue
from math import sqrt, pow, hypot, acos, degrees

class GraphNode:
    def __init__(self, x, y, cost=0, heuristic=0, prev=None):
        self.x = x
        self.y = y
        self.cost = cost
        self.heuristic = heuristic
        self.prev = prev
    
    def __lt__(self, other):
        return (self.cost + self.heuristic) < (other.cost + other.heuristic)

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y
    
    def __hash__(self):
        return hash((self.x, self.y))
    
    def __add__(self, other):
        return GraphNode(self.x + other[0], self.y + other[1])

class ThetaStarSmoothPlanner(Node):
    def __init__(self):
        super().__init__("theta_star_smooth_planner")

        self.declare_parameter("cost_limit", 20)
        self.cost_limit = self.get_parameter("cost_limit").value

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        map_qos = QoSProfile(depth=10)
        map_qos.durability = DurabilityPolicy.TRANSIENT_LOCAL

        self.map_sub = self.create_subscription(
            OccupancyGrid, "/costmap", self.map_callback, map_qos
        )
        self.pose_sub = self.create_subscription(
            PoseStamped, "/goal_pose", self.goal_callback, 10
        )
        self.path_pub = self.create_publisher(Path, "/theta_star/path", 10)
        self.map_pub = self.create_publisher(OccupancyGrid, "/theta_star/visited_map", 10)

        self.map_ = None
        self.visited_map_ = OccupancyGrid()
        self.map_resolution_ = None

        self.get_logger().info("theta_star_planner node has just started")

    def map_callback(self, map_msg: OccupancyGrid):
        self.map_ = map_msg
        self.visited_map_.header.frame_id = map_msg.header.frame_id
        self.visited_map_.info = map_msg.info
        self.visited_map_.data = [-1] * (map_msg.info.height * map_msg.info.width)
        self.map_resolution_ = map_msg.info.resolution

    def goal_callback(self, pose: PoseStamped):
        if self.map_ is None:
            self.get_logger().error("No map received!")
            return

        self.visited_map_.data = [-1] * (self.visited_map_.info.height * self.visited_map_.info.width)

        try:
            map_to_base_tf = self.tf_buffer.lookup_transform(
                self.map_.header.frame_id, "base_link", rclpy.time.Time()
            )
        except LookupException:
            self.get_logger().error("Could not transform from map to base_link")
            return

        map_to_base_pose = Pose()
        map_to_base_pose.position.x = map_to_base_tf.transform.translation.x
        map_to_base_pose.position.y = map_to_base_tf.transform.translation.y
        map_to_base_pose.orientation = map_to_base_tf.transform.rotation

        path = self.plan(map_to_base_pose, pose.pose)
        if path.poses:
            self.get_logger().info("Shortest path found!")
            self.get_logger().info(f'path_count = {len(path.poses)}')
            full_path = self.fill_up_path(path)
            self.get_logger().info(f'full_path_count = {len(full_path.poses)}')
            self.path_pub.publish(full_path)
        else:
            self.get_logger().warn("No path found to the goal.")

    def plan(self, start: Pose, goal: Pose):

        explore_directions = [
            (-1, 0), (1, 0), (0, -1), (0, 1),
            (-1, -1), (-1, 1), (1, -1), (1, 1)
        ]

        open_set = PriorityQueue()
        closed_set = set()
        g_score = {}

        start_node = self.world_to_grid(start)
        goal_node = self.world_to_grid(goal)

        start_node.cost = 0.0
        start_node.heuristic = self.euclidean_distance(start_node, goal_node)
        start_node.prev = start_node

        g_score[(start_node.x, start_node.y)] = 0.0
        open_set.put(start_node)

        goal_reached = False

        while not open_set.empty() and rclpy.ok():
            current = open_set.get()

            if (current.x, current.y) in closed_set:
                continue

            if current == goal_node:
                goal_node = current
                goal_reached = True
                break

            closed_set.add((current.x, current.y))

            for dx, dy in explore_directions:
                neighbor = current + (dx, dy)

                if not self.pose_on_map(neighbor):
                    continue

                if not (0 <= self.map_.data[self.pose_to_cell(neighbor)] < self.cost_limit):
                    continue

                if (neighbor.x, neighbor.y) in closed_set:
                    continue

                # -------- Theta* Core Logic --------
                if current.prev and self.line_of_sight(current.prev, neighbor):
                    parent = current.prev
                    new_cost = g_score[(parent.x, parent.y)] + self.euclidean_distance(parent, neighbor)
                else:
                    parent = current
                    new_cost = g_score[(current.x, current.y)] + self.euclidean_distance(current, neighbor)

                key = (neighbor.x, neighbor.y)

                if new_cost < g_score.get(key, float('inf')):
                    g_score[key] = new_cost
                    neighbor.cost = new_cost
                    neighbor.heuristic = self.euclidean_distance(neighbor, goal_node)
                    neighbor.prev = parent
                    open_set.put(neighbor)

            # Visualization (optional)
            # self.visited_map_.data[self.pose_to_cell(current)] = -106
            # self.map_pub.publish(self.visited_map_)

        # -------- Path Reconstruction --------
        path = Path()
        path.header.frame_id = self.map_.header.frame_id

        if not goal_reached:
            return path

        node = goal_node

        while rclpy.ok():
            pose = self.grid_to_world(node)
            pose_stamped = PoseStamped()
            pose_stamped.header.frame_id = self.map_.header.frame_id
            pose_stamped.pose = pose
            path.poses.append(pose_stamped)

            if node == node.prev:   # reached start
                break

            node = node.prev

        path.poses.reverse()
        return path

    def euclidean_distance(self, node: GraphNode, goal_node: GraphNode):
        return sqrt( pow(node.x - goal_node.x, 2) + pow(node.y - goal_node.y, 2) )
    
    def manhattan_distance(self, node: GraphNode, goal_node: GraphNode):
        return abs(node.x - goal_node.x) + abs(node.y - goal_node.y)

    def pose_on_map(self, node: GraphNode):
        return 0 <= node.x < self.map_.info.width and 0 <= node.y < self.map_.info.height

    def world_to_grid(self, pose: Pose) -> GraphNode:
        grid_x = int((pose.position.x - self.map_.info.origin.position.x) / self.map_.info.resolution)
        grid_y = int((pose.position.y - self.map_.info.origin.position.y) / self.map_.info.resolution)
        return GraphNode(grid_x, grid_y)

    def grid_to_world(self, node: GraphNode) -> Pose:
        pose = Pose()
        pose.position.x = node.x * self.map_.info.resolution + self.map_.info.origin.position.x
        pose.position.y = node.y * self.map_.info.resolution + self.map_.info.origin.position.y
        return pose

    def pose_to_cell(self, node: GraphNode):
        return node.y * self.map_.info.width + node.x
    
    def bresenham_line(self, start: GraphNode, end: GraphNode):
        x0, y0 = start.x, start.y
        x1, y1 = end.x, end.y

        dx = abs(x1 - x0)
        dy = abs(y1 - y0)

        sx = 1 if x0 < x1 else -1
        sy = 1 if y0 < y1 else -1

        err = dx - dy

        line = []

        while True:
            line.append(GraphNode(x0, y0))

            if x0 == x1 and y0 == y1:
                break

            e2 = 2 * err
            if e2 > -dy:
                err -= dy
                x0 += sx
            if e2 < dx:
                err += dx
                y0 += sy

        return line
    
    def line_crosses_obstacle(self, line):
        width = self.map_.info.width
        height = self.map_.info.height
        data = self.map_.data
        cost_limit = self.cost_limit

        for grid_pose in line:
            x = int(grid_pose.x)
            y = int(grid_pose.y)

            # Bounds check (CRITICAL)
            if not (0 <= x < width and 0 <= y < height):
                return True   # treat outside map as obstacle

            index = y * width + x

            if not (0 <= data[index] < cost_limit):
                return True

        return False
    
    def line_of_sight(self, start: GraphNode, end: GraphNode):
        line = self.bresenham_line(start, end)
        crosses = self.line_crosses_obstacle(line)
        return not crosses   # ← THIS is the fix
    
    def add_straight_line_poses(self, start_pose: PoseStamped, end_pose: PoseStamped, resolution):
        poses = []

        dx = end_pose.pose.position.x - start_pose.pose.position.x
        dy = end_pose.pose.position.y - start_pose.pose.position.y
        distance = hypot(dx, dy)

        if distance == 0:
            return []

        no_of_iter = int(distance / resolution)

        if no_of_iter == 0:
            return []

        x_increment = dx / no_of_iter
        y_increment = dy / no_of_iter

        for i in range(no_of_iter):
            pose = PoseStamped()   # ← NEW OBJECT EACH TIME
            pose.header.frame_id = start_pose.header.frame_id
            pose.pose.position.x = start_pose.pose.position.x + x_increment * i
            pose.pose.position.y = start_pose.pose.position.y + y_increment * i
            pose.pose.orientation = start_pose.pose.orientation
            poses.append(pose)

        # Add exact end pose
        poses.append(end_pose)

        return poses
    
    def fill_up_path(self, path: Path):

        filled_path = Path()
        filled_path.header.frame_id = path.header.frame_id

        if not path.poses:
            return filled_path

        pose_list = path.poses
        filled_path.poses.append(pose_list[0])

        for i in range(1, len(pose_list)):
            poses = self.add_straight_line_poses(
                pose_list[i - 1],
                pose_list[i],
                self.map_.info.resolution
            )
            filled_path.poses.extend(poses[1:])

        return filled_path


def main(args=None):
    rclpy.init(args=args)
    node = ThetaStarSmoothPlanner()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/theta_star_smooth_planner.cpp
#include <rclcpp/rclcpp.hpp>

#include <nav_msgs/msg/occupancy_grid.hpp>
#include <nav_msgs/msg/path.hpp>

#include <geometry_msgs/msg/pose_stamped.hpp>
#include <geometry_msgs/msg/pose.hpp>

#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>
#include <tf2/exceptions.h>

#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <cmath>
#include <algorithm>

using std::placeholders::_1;

// ======================================================
// GRAPH NODE (PYTHON-EQUIVALENT)
// ======================================================
struct GraphNode
{
    int x, y;

    double cost = 0;
    double heuristic = 0;

    GraphNode* prev = nullptr;

    GraphNode(int _x = 0, int _y = 0)
        : x(_x), y(_y) {}
};

// ======================================================
// PRIORITY QUEUE
// ======================================================
struct CompareNode
{
    bool operator()(GraphNode* a, GraphNode* b)
    {
        return (a->cost + a->heuristic) > (b->cost + b->heuristic);
    }
};

// ======================================================
// THETA STAR NODE
// ======================================================
class ThetaStarPlanner : public rclcpp::Node
{
public:
    ThetaStarPlanner()
        : Node("theta_star_planner"),
          tf_buffer_(this->get_clock()),
          tf_listener_(tf_buffer_)
    {
        this->declare_parameter("cost_limit", 20);
        cost_limit_ = this->get_parameter("cost_limit").as_int();

        map_sub_ = this->create_subscription<nav_msgs::msg::OccupancyGrid>(
            "/costmap", 10,
            std::bind(&ThetaStarPlanner::mapCallback, this, _1));

        goal_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
            "/goal_pose", 10,
            std::bind(&ThetaStarPlanner::goalCallback, this, _1));

        path_pub_ = this->create_publisher<nav_msgs::msg::Path>(
            "/theta_star/path", 10);

        RCLCPP_INFO(this->get_logger(), "Theta* (GraphNode version) ready");
    }

private:
    // ======================================================
    // VARIABLES
    // ======================================================
    nav_msgs::msg::OccupancyGrid map_;
    double resolution_;
    int cost_limit_;

    rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr map_sub_;
    rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr goal_sub_;
    rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr path_pub_;

    tf2_ros::Buffer tf_buffer_;
    tf2_ros::TransformListener tf_listener_;

    // ======================================================
    // MAP CALLBACK
    // ======================================================
    void mapCallback(const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
    {
        map_ = *msg;
        resolution_ = msg->info.resolution;
    }

    // ======================================================
    // GOAL CALLBACK
    // ======================================================
    void goalCallback(const geometry_msgs::msg::PoseStamped::SharedPtr goal)
    {
        if (map_.data.empty())
            return;

        geometry_msgs::msg::Pose start;

        try
        {
            auto tf = tf_buffer_.lookupTransform(
                map_.header.frame_id, "base_link", tf2::TimePointZero);

            start.position.x = tf.transform.translation.x;
            start.position.y = tf.transform.translation.y;
            start.orientation = tf.transform.rotation;
        }
        catch (tf2::TransformException &ex)
        {
            RCLCPP_ERROR(this->get_logger(), "%s", ex.what());
            return;
        }

        auto path = plan(start, goal->pose);
        path_pub_->publish(fillUpPath(path));
    }

    // ======================================================
    // PLAN (THETA*)
    // ======================================================
    nav_msgs::msg::Path plan(const geometry_msgs::msg::Pose &start,
                              const geometry_msgs::msg::Pose &goal)
    {
        std::priority_queue<GraphNode*, std::vector<GraphNode*>, CompareNode> open;
        std::unordered_set<long long> closed;
        std::unordered_map<long long, double> g_score;

        auto start_node = new GraphNode(worldToGrid(start));
        auto goal_node  = new GraphNode(worldToGrid(goal));

        start_node->cost = 0;
        start_node->heuristic = heuristic(*start_node, *goal_node);
        start_node->prev = start_node;

        open.push(start_node);
        g_score[key(start_node->x, start_node->y)] = 0;

        std::vector<std::pair<int,int>> dirs = {
            {-1,0},{1,0},{0,-1},{0,1},
            {-1,-1},{-1,1},{1,-1},{1,1}
        };

        GraphNode* final = nullptr;
        bool found = false;

        while (!open.empty())
        {
            GraphNode* current = open.top();
            open.pop();

            if (closed.count(key(current->x, current->y)))
                continue;

            if (current->x == goal_node->x &&
                current->y == goal_node->y)
            {
                final = current;
                found = true;
                break;
            }

            closed.insert(key(current->x, current->y));

            for (auto d : dirs)
            {
                int nx = current->x + d.first;
                int ny = current->y + d.second;

                if (!inBounds(nx, ny))
                    continue;

                if (map_.data[toIndex(nx, ny)] >= cost_limit_)
                    continue;

                auto neighbor = new GraphNode(nx, ny);

                GraphNode* parent = current;

                if (current->prev &&
                    lineOfSight(current->prev->x, current->prev->y, nx, ny))
                {
                    parent = current->prev;
                }

                double new_cost = getG(parent) +
                                  heuristic(*parent, *neighbor);

                long long k = key(nx, ny);

                if (!g_score.count(k) || new_cost < g_score[k])
                {
                    g_score[k] = new_cost;

                    neighbor->cost = new_cost;
                    neighbor->heuristic = heuristic(*neighbor, *goal_node);
                    neighbor->prev = parent;

                    open.push(neighbor);
                }
            }
        }

        nav_msgs::msg::Path path;
        path.header.frame_id = map_.header.frame_id;

        if (!found)
            return path;

        GraphNode* node = final;

        while (node)
        {
            geometry_msgs::msg::PoseStamped ps;
            ps.header.frame_id = map_.header.frame_id;
            ps.pose = gridToWorld(*node);

            path.poses.push_back(ps);

            if (node->prev == node)
                break;

            node = node->prev;
        }

        std::reverse(path.poses.begin(), path.poses.end());
        return path;
    }

    // ======================================================
    // HELPERS
    // ======================================================
    double heuristic(GraphNode a, GraphNode b)
    {
        return std::hypot(a.x - b.x, a.y - b.y);
    }

    double getG(GraphNode* n)
    {
        return n ? n->cost : 0;
    }

    long long key(int x, int y)
    {
        return ((long long)x << 32) | y;
    }

    bool inBounds(int x, int y)
    {
        return x >= 0 && y >= 0 &&
               x < (int)map_.info.width &&
               y < (int)map_.info.height;
    }

    int toIndex(int x, int y)
    {
        return y * map_.info.width + x;
    }

    GraphNode worldToGrid(const geometry_msgs::msg::Pose &p)
    {
        int x = (p.position.x - map_.info.origin.position.x) / resolution_;
        int y = (p.position.y - map_.info.origin.position.y) / resolution_;
        return GraphNode(x, y);
    }

    geometry_msgs::msg::Pose gridToWorld(const GraphNode &n)
    {
        geometry_msgs::msg::Pose p;
        p.position.x = n.x * resolution_ + map_.info.origin.position.x;
        p.position.y = n.y * resolution_ + map_.info.origin.position.y;
        return p;
    }

    // ======================================================
    // LINE OF SIGHT
    // ======================================================
    bool lineOfSight(int x0, int y0, int x1, int y1)
    {
        int dx = abs(x1 - x0), dy = abs(y1 - y0);
        int sx = x0 < x1 ? 1 : -1;
        int sy = y0 < y1 ? 1 : -1;
        int err = dx - dy;

        while (true)
        {
            if (map_.data[toIndex(x0, y0)] >= cost_limit_)
                return false;

            if (x0 == x1 && y0 == y1)
                break;

            int e2 = 2 * err;

            if (e2 > -dy) { err -= dy; x0 += sx; }
            if (e2 < dx)  { err += dx; y0 += sy; }
        }

        return true;
    }

    // ======================================================
    // ADD STRAIGHT LINE (PYTHON MATCH)
    // ======================================================
    std::vector<geometry_msgs::msg::PoseStamped>
    addStraightLinePoses(const geometry_msgs::msg::PoseStamped &a,
                         const geometry_msgs::msg::PoseStamped &b,
                         double res)
    {
        std::vector<geometry_msgs::msg::PoseStamped> out;

        double dx = b.pose.position.x - a.pose.position.x;
        double dy = b.pose.position.y - a.pose.position.y;

        double dist = std::hypot(dx, dy);
        if (dist == 0) return out;

        int steps = (int)(dist / res);
        if (steps <= 0) return out;

        double ix = dx / steps;
        double iy = dy / steps;

        for (int i = 0; i < steps; i++)
        {
            geometry_msgs::msg::PoseStamped p;
            p.header.frame_id = a.header.frame_id;
            p.pose.position.x = a.pose.position.x + ix * i;
            p.pose.position.y = a.pose.position.y + iy * i;
            p.pose.orientation = a.pose.orientation;
            out.push_back(p);
        }

        out.push_back(b);
        return out;
    }

    // ======================================================
    // FILL UP PATH (PYTHON MATCH)
    // ======================================================
    nav_msgs::msg::Path fillUpPath(const nav_msgs::msg::Path &path)
    {
        nav_msgs::msg::Path filled;
        filled.header.frame_id = path.header.frame_id;

        if (path.poses.empty())
            return filled;

        filled.poses.push_back(path.poses[0]);

        for (size_t i = 1; i < path.poses.size(); i++)
        {
            auto seg = addStraightLinePoses(
                path.poses[i - 1],
                path.poses[i],
                resolution_);

            for (size_t j = 1; j < seg.size(); j++)
                filled.poses.push_back(seg[j]);
        }

        return filled;
    }
};

// ======================================================
// MAIN
// ======================================================
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ThetaStarPlanner>());
    rclcpp::shutdown();
    return 0;
}